app/models/course/enrol_request.rb

```diff
@@ -0,0 +1,25 @@
+class Course::EnrolRequest < ActiveRecord::Base
+  stampable
+  has_paper_trail on: [:destroy]
+  validates :course_id, uniqueness: { scope: :user_id }
+
+  scope :student, -> { where(role: CourseUser.roles[:student]) }
+  scope :staff, -> { where(role: [CourseUser.roles[:teaching_assistant],
+                                  CourseUser.roles[:manager],
+                                  CourseUser.roles[:owner]]) }
+
+  enum role: CourseUser.roles
+
+  belongs_to :course
+  belongs_to :user
+
+  def approve!
+    unless CourseUser.where(course_id: course_id, user_id: user_id).empty?
+      return
```

Comment:
this function is racy. approve can still error when more than one user approves the request at the same time.